File: .github/scripts/apply-pr-labels.py
import json
import os
import re
import subprocess
import sys

VALID_TYPES = {"feature", "fix", "chore", "docs", "refactor", "test"}
VALID_AREAS = {"membership", "admin", "training", "auth", "site", "atc", "roster", "api", "integrations", "infrastructure"}
VALID_MISSING = {"tests", "copy", "docs", "migration"}

output_file = os.environ.get("OPENCODE_OUTPUT")
pr_number = os.environ.get("PR_NUMBER")

if not output_file or not pr_number:
    print("Missing required env vars: OPENCODE_OUTPUT, PR_NUMBER", file=sys.stderr)
    sys.exit(1)


def gh(args: str, stdin: str | None = None) -> str:
    return subprocess.run(
        f"gh {args}",
        shell=True,
        capture_output=True,
        text=True,
        input=stdin,
        env={**os.environ, "NO_COLOR": "1"},
        check=True,
    ).stdout.strip()


def extract_json(text: str) -> str:
    text = re.sub(r"```(?:json)?\s*", "", text)
    text = re.sub(r"```\s*", "", text).strip()

    start = text.find("{")
    if start == -1:
        return text

    depth = 0
    for i, ch in enumerate(text[start:], start):
        if ch == "{":
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0:
                return text[start : i + 1]

    return text


def validate(classification: dict) -> dict:
    t = classification.get("type", "")
    if t not in VALID_TYPES:
        raise ValueError(f'Invalid type: "{t}". Must be one of: {", ".join(sorted(VALID_TYPES))}')

    areas = classification.get("area", [])
    if not isinstance(areas, list):
        raise ValueError("'area' must be an array")

    for item in areas:
        if item not in VALID_AREAS:
            raise ValueError(f'Invalid area: "{item}". Must be one of: {", ".join(sorted(VALID_AREAS))}')

    missing = classification.get("missing", [])
    if not isinstance(missing, list):
        raise ValueError("'missing' must be an array")

    for item in missing:
        if item not in VALID_MISSING:
            raise ValueError(f'Invalid missing item: "{item}". Must be one of: {", ".join(sorted(VALID_MISSING))}')

    return {
        "type": t,
        "area": list(set(areas)),
        "missing": list(set(missing)),
        "reasoning": classification.get("reasoning"),
    }


try:
    with open(output_file) as f:
        raw = f.read().strip()

    json_str = extract_json(raw)
    print("Extracted:", json_str[:200])

    classification = validate(json.loads(json_str))
    print("Classification:", json.dumps(classification))

    desired = {f"type:{classification['type']}"}
    desired |= {f"area:{item}" for item in classification["area"]}
    desired |= {f"needs-{item}" for item in classification["missing"]}
    print("Desired labels:", desired)

    current = set(gh(f"pr view {pr_number} --json labels --jq '.labels[].name'").split("\n"))
    current.discard("")
    print("Current labels:", current)

    managed_prefixes = ("type:", "area:", "needs-")

    for label in current:
        if label.startswith(managed_prefixes) and label not in desired:
            print(f"Removing: {label}")
            gh(f'pr edit {pr_number} --remove-label "{label}"')

    for label in desired:
        if label not in current:
            print(f"Adding: {label}")
            gh(f'pr edit {pr_number} --add-label "{label}"')

    print("Labels synced successfully")

except (ValueError, json.JSONDecodeError) as e:
    print(f"Classification failed: {e}", file=sys.stderr)
    sys.exit(1)

File: .github/scripts/apply-pr-labels.js
const fs = require("node:fs");
const path = require("node:path");
const { execSync } = require("node:child_process");

const VALID_TYPES = ["feature", "fix", "chore", "docs", "refactor", "test"];
const VALID_MISSING = ["tests", "copy", "docs", "migration"];

const outputFile = process.env.OPENCODE_OUTPUT;
const prNumber = process.env.PR_NUMBER;

if (!outputFile || !prNumber) {
  console.error("Missing required env vars: OPENCODE_OUTPUT, PR_NUMBER");
  process.exit(1);
}

function extractModelResponse(filePath) {
  const content = fs.readFileSync(filePath, "utf-8");
  const lines = content.trim().split("\n");

  const lastLine = lines[lines.length - 1];
  try {
    const event = JSON.parse(lastLine);
    if (event.type === "assistant" && typeof event.text === "string") {
      return event.text.trim();
    }
  } catch {
    // not JSON, treat whole content as text
  }

  // Fallback: try to find JSON in the raw output
  const match = content.match(/\{[\s\S]*"type"[\s\S]*"missing"[\s\S]*\}/);
  if (match) return match[0].trim();

  return content.trim();
}

function validateClassification(obj) {
  if (!obj || typeof obj !== "object") {
    throw new Error("Classification is not an object");
  }

  if (!VALID_TYPES.includes(obj.type)) {
    throw new Error(
      `Invalid type: "${obj.type}". Must be one of: ${VALID_TYPES.join(", ")}`
    );
  }

  if (!Array.isArray(obj.missing)) {
    throw new Error("'missing' must be an array");
  }

  for (const item of obj.missing) {
    if (!VALID_MISSING.includes(item)) {
      throw new Error(
        `Invalid missing item: "${item}". Must be one of: ${VALID_MISSING.join(", ")}`
      );
    }
  }

  obj.missing = [...new Set(obj.missing)];

  return {
    type: obj.type,
    missing: obj.missing,
    reasoning: typeof obj.reasoning === "string" ? obj.reasoning : null,
  };
}

function gh(args, stdin) {
  return execSync(`gh ${args}`, {
    encoding: "utf-8",
    input: stdin || undefined,
    env: { ...process.env, NO_COLOR: "1" },
  }).trim();
}

try {
  const rawText = extractModelResponse(outputFile);
  console.log("Extracted response:", rawText.slice(0, 200));

  const cleaned = rawText
    .replace(/```json\s*/g, "")
    .replace(/```\s*/g, "")
    .trim();

  const parsed = JSON.parse(cleaned);
  const classification = validateClassification(parsed);
  console.log("Classification:", JSON.stringify(classification));

  const desiredLabels = new Set([`type:${classification.type}`]);
  for (const item of classification.missing) {
    desiredLabels.add(`needs-${item}`);
  }
  console.log("Desired labels:", [...desiredLabels]);

  const currentLabelsJson = gh(
    `pr view ${prNumber} --json labels --jq '.labels[].name'`
  );
  const currentLabels = new Set(
    currentLabelsJson.split("\n").filter(Boolean)
  );
  console.log("Current labels:", [...currentLabels]);

  const managedPrefixes = ["type:", "needs-"];

  for (const label of currentLabels) {
    if (
      managedPrefixes.some((p) => label.startsWith(p)) &&
      !desiredLabels.has(label)
    ) {
      console.log(`Removing label: ${label}`);
      gh(`pr edit ${prNumber} --remove-label "${label}"`);
    }
  }

  for (const label of desiredLabels) {
    if (!currentLabels.has(label)) {
      console.log(`Adding label: ${label}`);
      gh(`pr edit ${prNumber} --add-label "${label}"`);
    }
  }

  console.log("Labels synced successfully");
} catch (err) {
  console.error("Classification failed:", err.message);
  if (err.message.includes("Invalid ") || err instanceof SyntaxError) {
    console.error(
      "Model returned unparseable or invalid output. No labels applied."
    );
  }
  process.exit(1);
}
